fdbserver/DataDistribution.actor.cpp

```diff
@@ -2028,6 +2272,140 @@ struct DDTeamCollection : ReferenceCounted<DDTeamCollection> {
+		}
+		wait(delay(FLOW_KNOBS->PREVENT_FAST_SPIN_DELAY, TaskLowPriority)); //After the team trackers wait on the initial failure reaction delay, they yield. We want to make sure every tracker has had the opportunity to send their relocations to the queue.
+		if(!self->zeroHealthyTeams->get() && !self->processingUnhealthy->get()) {
+			break;
+		}
+	}
+	wait(self->addSubsetComplete.getFuture());
+	TraceEvent("DDRemovingBadTeams", self->masterId).detail("Primary", self->primary);
+	for(auto it : self->badTeams) {
+		it->tracker.cancel();
+	}
+	self->badTeams.clear();
+	return Void();
+}
+
+ACTOR Future<Void> teamRemover(DDTeamCollection* self) {
+	state int numMachineTeamRemoved = 0;
+	loop {
+		// Wait on processingUnhealthy as removeBadTeams() does
+		loop {
+			while(self->zeroHealthyTeams->get() || self->processingUnhealthy->get()) {
+				wait(self->zeroHealthyTeams->onChange() || self->processingUnhealthy->onChange());
+			}
+			wait(delay(FLOW_KNOBS->PREVENT_FAST_SPIN_DELAY, TaskLowPriority)); //After the team trackers wait on the initial failure reaction delay, they yield. We want to make sure every tracker has had the opportunity to send their relocations to the queue.
+			if(!self->zeroHealthyTeams->get() && !self->processingUnhealthy->get()) {
+				break;
+			}
+		}
+
+		// loop until the number of machine teams <= desiredMachineTeams
+		int totalHealthyMachineCount = self->calculateHealthyMachineCount();
+		int currentHealthyMTCount = self->getHealthyMachineTeamCount();
+
+		int desiredMachineTeams = SERVER_KNOBS->DESIRED_TEAMS_PER_MACHINE * totalHealthyMachineCount;
+		int maxMachineTeams = SERVER_KNOBS->MAX_TEAMS_PER_MACHINE * totalHealthyMachineCount;
+
+		if (currentHealthyMTCount > desiredMachineTeams || self->machineTeams.size() > maxMachineTeams) {
+			// Pick the machine team with the least number of process teams and mark in undesired
+			state Reference<TCMachineTeamInfo> mt;
+			// Q: Should we prefer to remove an unhealthy machine team?
+			// i.e., if the currentHealthyMTCount < self->machineTeams.size(), find an unhealthy machine team to remove
+			state int minNumProcessTeams = self->getMachineTeamWithLeastProcessTeams(mt);
+			ASSERT(mt.isValid());
+			mt->redundant = true;  // unused as a placeholder to mark the machine team status
+
+			// Pick one process team, and mark it as a bad team
+			// Remove the machine by removing its process team one by one
+			state bool addedNewBadTeam = false;
+			state Reference<TCTeamInfo> team;
+			state int teamIndex = 0;
+			state int numServerTeamRemoved = 0;
+			for (teamIndex = 0; teamIndex < self->teams.size(); ++teamIndex) {
+				team = self->teams[teamIndex];
+				if (team->machineTeam->machineIDs != mt->machineIDs) {
+					continue;
+				}
+				// The team will be marked as a bad team
+				team->redundant = true; // unused as a placeholder to mark the team status
+				numServerTeamRemoved++;
+				if( self->removeTeam(team) ) {
```

Comment:
Should always return true because we just found the team. May change it to ASSERT.